src/Juvix/Compiler/Core/Transformation/Eta.hs

```diff
@@ -33,13 +33,13 @@ etaExpandConstrs argsNum = umap go
 squashApps :: Node -> Node
 squashApps = dmap go
   where
-    go :: Node -> Node
+    go :: Node -> Maybe Node
     go n =
       let (l, args) = unfoldApp n
        in case l of
-            Constr i tag args' -> Constr i tag (args' ++ args)
-            BuiltinApp i op args' -> BuiltinApp i op (args' ++ args)
-            _ -> n
+            Constr i tag args' -> Just (Constr i tag (map squashApps (args' ++ args)))
```

Comment:
This is wrong, because it stops on the first Constr it sees top-down. It should recursively go into the arguments of Constr not to miss some Constr there.

Please restore the old dmap and its variants, and name the new ones dmap', dmapM', etc.

Reply:
Oh, sorry I didn't notice the recursive invocation of squashApps

Reply:
But I still think old dmap variants should be restored. The point is that the *N and *B variants manage the binder information for you, but you need to do this manually when you do the recursive call yourself.